streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -761,7 +761,7 @@ private void runLoop() {
                     "Will close out all assigned tasks and rejoin the consumer group.");
 
                 taskManager.handleLostAll();
-                enforceRebalance();
+                needRebalance = true;
```

Comment:
Why not trigger rebalance immediately but defer it to the next iteration? Is it to make sure we capture any exceptions from the `enforceRebalance`? I thought this function should never throw any new exception since we already close all tasks as lost.

Reply:
Also cc @ableegoldman who's working on the new consumer API to replace enforceRebalance.

Reply:
John and I were discussing whether to remove the call to `handleLostAll` here :)

Reply:
Also +1 on this -- just call `enforceRebalance` directly here (and remove the `handleLostAll`)

Reply:
Actually do we even need the `needRebalance` flag? For version probing we just call `enforceRebalance` directly, and for `TaskMigratedException` it seems we should do the same. Maybe at some point in the future we will need to introduce it again, but for now it seems like we can get away with no `rebalanceException` AND no `needRebalance` flag